daemon/cluster/executor/container/controller.go

```diff
@@ -153,21 +154,69 @@ func (r *controller) Wait(pctx context.Context) error {
 	ctx, cancel := context.WithCancel(pctx)
 	defer cancel()
 
-	err := r.adapter.wait(ctx)
-	if ctx.Err() != nil {
-		return ctx.Err()
-	}
-	if err != nil {
-		ee := &exitError{}
-		if err.Error() != "" {
-			ee.cause = err
+	stop := make(chan struct{})
+	errChan := make(chan error)
+
+	go func() {
+		err := r.adapter.wait(ctx)
+		select {
+		case <-stop:
+		case errChan <- err:
 		}
-		if ec, ok := err.(exec.ExitCoder); ok {
-			ee.code = ec.ExitCode()
+	}()
+
+	go func() {
+		// polling health status every 3 seconds
```

Comment:
Probing health status doesn't sound right to me. Engine should emit events on container healthiness. Cluster manager should consume the events.